src/app/src/utils/providers/gitlab.ts

```diff
@@ -0,0 +1,168 @@
+      const fileMetadata = await $api(`/repository/files/${encodedPath}?ref=${branch}`)
+
+      const gitFile: GitFile = {
+        name: path.split('/').pop() || path,
+        path,
+        sha: fileMetadata.blob_id,
+        size: fileMetadata.size,
+        url: fileMetadata.file_path,
+        content: fileMetadata.content,
+        encoding: 'base64' as const,
+        provider: 'gitlab' as const,
+      }
+
+      if (cached) {
+        gitFiles[path] = gitFile
+      }
+      return gitFile
+    }
+    catch (error) {
+      // Handle different types of errors gracefully
+      if ((error as { status?: number }).status === 404) {
+        console.warn(`File not found on GitLab: ${path}`)
+        return null
+      }
+
+      console.error(`Failed to fetch file from GitLab: ${path}`, error)
+
+      // For development, show alert. In production, you might want to use a toast notification
+      if (process.env.NODE_ENV === 'development') {
+        alert(`Failed to fetch file: ${path}\n${(error as { message?: string }).message || error}`)
+      }
+
+      return null
+    }
+  }
+
+  function commitFiles(files: RawFile[], message: string): Promise<CommitResult | null> {
+    if (!token) {
+      return Promise.resolve(null)
+    }
+
+    files = files
+      .filter(file => file.status !== DraftStatus.Pristine)
+      .map(file => ({ ...file, path: joinURL(rootDir, file.path) }))
+
+    return commitFilesToGitLab({
+      owner,
+      repo,
+      branch,
+      files,
+      message,
+      authorName,
+      authorEmail,
+    })
+  }
+
+  async function commitFilesToGitLab({ branch, files, message, authorName, authorEmail }: CommitFilesOptions) {
+    // GitLab uses a single commits API with actions
+    const actions = files.map((file) => {
+      if (file.status === DraftStatus.Deleted) {
+        return {
+          action: 'delete',
+          file_path: file.path,
+        }
+      }
+      else if (file.status === DraftStatus.Created) {
+        return {
+          action: 'create',
+          file_path: file.path,
+          content: file.content,
+          encoding: file.encoding === 'base64' ? 'base64' : 'text',
+        }
+      }
+      else {
+        return {
+          action: 'update',
+          file_path: file.path,
+          content: file.content,
+          encoding: file.encoding === 'base64' ? 'base64' : 'text',
+        }
+      }
+    })
+
+    const commitData = await $api(`/repository/commits`, {
+      method: 'POST',
+      body: {
+        branch,
+        commit_message: message,
+        actions,
+        author_name: authorName,
+        author_email: authorEmail,
+      },
+    })
+
+    return {
+      success: true,
+      commitSha: commitData.id,
+      url: `${instanceUrl}/${owner}/${repo}/-/commit/${commitData.id}`,
+    }
+  }
+
+  function getRepositoryUrl() {
+    return `${instanceUrl}/${owner}/${repo}`
```

Comment:
With GitLab, we can have multiple nested groups. This leads to the URL `${instanceUrl}/${group}/${group2}/${group3}/${repo}`, so should we pass the owner as `group/group2/group3`? Because we can access:

<img width="749" height="213" alt="Image" src="https://github.com/user-attachments/assets/a0ba61b6-c217-4c62-9391-c9874ffd31d6" />

(a question, not a request for change)

Reply:
Seems easier to me to handle it this way indeed